signal/signal.go

```diff
@@ -19,8 +22,78 @@ var (
 	started int32
 )
 
+// systemdNotifyReady notifies systemd about LND being ready, logs the result of
+// the operation or possible error. Besides logging, systemd being unavailable
+// is ignored.
+func systemdNotifyReady() error {
+	notified, err := daemon.SdNotify(false, daemon.SdNotifyReady)
+	if err != nil {
+		err := fmt.Errorf("failed to notify systemd %v (if you aren't "+
+			"running systemd clear the environment variable "+
+			"NOTIFY_SOCKET)", err)
+		log.Error(err)
+
+		// The SdNotify doc says it's common to ignore the
+		// error. We don't want to ignore it because if someone
+		// set up systemd to wait for initialization other
+		// processes would get stuck.
+		return err
+	}
+	if notified {
+		log.Info("Systemd was notified about our readiness")
+	} else {
+		log.Info("We're not running within systemd")
+	}
+	return nil
+}
+
+// systemdNotifyStop notifies systemd that LND is stopping and logs error if
+// the notification failed. It also logs if the notification was actually sent.
+// Systemd being unavailable is intentionally ignored.
+func systemdNotifyStop() {
+	notified, err := daemon.SdNotify(false, daemon.SdNotifyStopping)
+
+	// Just log - we're stopping anyway.
+	if err != nil {
+		log.Errorf("Failed to notify systemd: %v", err)
+	}
+	if notified {
+		log.Infof("Systemd was notified about stopping")
+	}
+}
+
+// Notifier handles notifications about status of LND.
+type Notifier struct {
+	// notifiedReady remembers whether Ready was sent to avoid sending it
+	// multiple times.
+	notifiedReady bool
+}
+
+// NotifyReady notifies other applications that RPC is ready.
+func (notifier *Notifier) NotifyReady(walletUnlocked bool) error {
+	if !notifier.notifiedReady {
+		err := systemdNotifyReady()
+		if err != nil {
+			return err
+		}
+		notifier.notifiedReady = true
+	}
+	if walletUnlocked {
+		_, _ = daemon.SdNotify(false, "STATUS=Wallet unlocked")
+	} else {
+		_, _ = daemon.SdNotify(false, "STATUS=Wallet locked")
+	}
+
+	return nil
+}
+
+// NotifyStop notifies other applications that LND is stopping.
+func (notifier *Notifier) NotifyStop() {
```

Comment:
Hmm, indeed it's not *required* to be exported in the code as it's structured *now*. It could be reasonably considered a public interface that can be ready to use in the future. So not sure to unexport it or not. :)

Reply:
Generally I think you can just unexport and then we can export later if somebody needs it, it's a very easy refactor to export :) Only if you're going to push again, otherwise happy as is!

Reply:
OK, if I understand you correctly it's not hard requirement. I don't see anything else to do with the code and I feel like changing it and triggering re-review of two people is not worth it. Please LMK if I got this wrong.

Reply:
Yep, non-blocking.